src/com/toopher/openam/ToopherSecondFactor.java

```diff
@@ -73,154 +68,80 @@ private void debug_message(String message) {
 
     @Override
     public void init(Subject subject, Map sharedState, Map options) {
-        super.init(subject, sharedState, options);
+        if (isUseFirstPassEnabled()) {
+            this.sharedState = sharedState;
+        }
+        if (sharedState == null) {
+            debug_message("sharedState is null!");
+        }
+        if (options == null) {
+            debug_message("options is null");
+        }
+        this.options = options;
+
+        String authLevel = CollectionHelper.getMapAttr(options, KEY_AUTHLEVEL);
+        if (authLevel != null) {
+            try {
+                setAuthLevel(Integer.parseInt(authLevel));
+            } catch (Exception e) {
+                debug.error("ToopherSecondFactor.init() : " + "Unable to set auth level " + authLevel, e);
+            }
+        }
+
+        String toopherConsumerKey = CollectionHelper.getMapAttr(options, KEY_CONSUMER_KEY);
+        String toopherConsumerSecret = CollectionHelper.getMapAttr(options, KEY_CONSUMER_SECRET);
+        String toopherApiUrl = CollectionHelper.getMapAttr(options, KEY_API_URL);
+
+        api = new ToopherIframe(toopherConsumerKey, toopherConsumerSecret, toopherApiUrl);
+
+        userName = (String) sharedState.get(getUserKey());
+
+        String userEmailAttr = CollectionHelper.getMapAttr(options, KEY_MAIL_ATTR);
+        // TODO: figure out how to get the user's mail attribute into userEmail
+
+        requestToken = new BigInteger(20 * 8, secureRandom).toString(32);
     }
 
-    private int ajaxPollingResponse(boolean poll) throws JSONException, IOException {
-        HttpServletResponse response = getHttpServletResponse();
-        Map<String, Object> params = new HashMap<String, Object>();
-        params.put("poll", poll);
-        JSONObject json = new JSONObject(params);
-        json.write(response.getWriter());
-        response.getWriter().flush();
-        return 0;
+    private Map<String, String[]> decodeUrlEncodedDict(String urlEncodedDict) throws Exception {
+        String[] iframeDataStrs = urlEncodedDict.split("&");
+        Map<String, String[]> params = new HashMap<String, String[]>(iframeDataStrs.length);
```

Comment:
That's what `validatePostback` expects in its current incarnation.  As we evolve that function signature, this will change as well, but meanwhile...